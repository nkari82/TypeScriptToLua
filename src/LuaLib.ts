import * as path from "path";
import { EmitHost } from "./transpilation";
import * as lua from "./LuaAST";

export enum LuaLibFeature {
    ArrayConcat = "ArrayConcat",
    ArrayEntries = "ArrayEntries",
    ArrayEvery = "ArrayEvery",
    ArrayFilter = "ArrayFilter",
    ArrayForEach = "ArrayForEach",
    ArrayFind = "ArrayFind",
    ArrayFindIndex = "ArrayFindIndex",
    ArrayFrom = "ArrayFrom",
    ArrayIncludes = "ArrayIncludes",
    ArrayIndexOf = "ArrayIndexOf",
    ArrayIsArray = "ArrayIsArray",
    ArrayJoin = "ArrayJoin",
    ArrayMap = "ArrayMap",
    ArrayPush = "ArrayPush",
    ArrayPushArray = "ArrayPushArray",
    ArrayReduce = "ArrayReduce",
    ArrayReduceRight = "ArrayReduceRight",
    ArrayReverse = "ArrayReverse",
    ArrayUnshift = "ArrayUnshift",
    ArraySort = "ArraySort",
    ArraySlice = "ArraySlice",
    ArraySome = "ArraySome",
    ArraySplice = "ArraySplice",
    ArrayToObject = "ArrayToObject",
    ArrayFlat = "ArrayFlat",
    ArrayFlatMap = "ArrayFlatMap",
    ArraySetLength = "ArraySetLength",
    Await = "Await",
    Class = "Class",
    ClassExtends = "ClassExtends",
    CloneDescriptor = "CloneDescriptor",
    Decorate = "Decorate",
    DecorateParam = "DecorateParam",
    Delete = "Delete",
    DelegatedYield = "DelegatedYield",
    Error = "Error",
    FunctionBind = "FunctionBind",
    Generator = "Generator",
    InstanceOf = "InstanceOf",
    InstanceOfObject = "InstanceOfObject",
    Iterator = "Iterator",
    LuaIteratorSpread = "LuaIteratorSpread",
    Map = "Map",
    MathAtan2 = "MathAtan2",
    MathSign = "MathSign",
    New = "New",
    Number = "Number",
    NumberIsFinite = "NumberIsFinite",
    NumberIsNaN = "NumberIsNaN",
    NumberToString = "NumberToString",
    ObjectAssign = "ObjectAssign",
    ObjectDefineProperty = "ObjectDefineProperty",
    ObjectEntries = "ObjectEntries",
    ObjectFromEntries = "ObjectFromEntries",
    ObjectGetOwnPropertyDescriptor = "ObjectGetOwnPropertyDescriptor",
    ObjectGetOwnPropertyDescriptors = "ObjectGetOwnPropertyDescriptors",
    ObjectKeys = "ObjectKeys",
    ObjectRest = "ObjectRest",
    ObjectValues = "ObjectValues",
    ParseFloat = "ParseFloat",
    ParseInt = "ParseInt",
    Promise = "Promise",
    PromiseAll = "PromiseAll",
    PromiseAllSettled = "PromiseAllSettled",
    PromiseAny = "PromiseAny",
    PromiseRace = "PromiseRace",
    Set = "Set",
    SetDescriptor = "SetDescriptor",
    SparseArrayNew = "SparseArrayNew",
    SparseArrayPush = "SparseArrayPush",
    SparseArraySpread = "SparseArraySpread",
    WeakMap = "WeakMap",
    WeakSet = "WeakSet",
    SourceMapTraceBack = "SourceMapTraceBack",
    Spread = "Spread",
    StringAccess = "StringAccess",
    StringCharAt = "StringCharAt",
    StringCharCodeAt = "StringCharCodeAt",
    StringEndsWith = "StringEndsWith",
    StringIncludes = "StringIncludes",
    StringPadEnd = "StringPadEnd",
    StringPadStart = "StringPadStart",
    StringReplace = "StringReplace",
    StringReplaceAll = "StringReplaceAll",
    StringSlice = "StringSlice",
    StringSplit = "StringSplit",
    StringStartsWith = "StringStartsWith",
    StringSubstr = "StringSubstr",
    StringSubstring = "StringSubstring",
    StringTrim = "StringTrim",
    StringTrimEnd = "StringTrimEnd",
    StringTrimStart = "StringTrimStart",
    Symbol = "Symbol",
    SymbolRegistry = "SymbolRegistry",
    TypeOf = "TypeOf",
    Unpack = "Unpack",
}

export interface LuaLibFeatureInfo {
    dependencies?: LuaLibFeature[];
    exports: string[];
}
export type LuaLibModulesInfo = Record<LuaLibFeature, LuaLibFeatureInfo>;

export const luaLibModulesInfoFileName = "lualib_module_info.json";
let luaLibModulesInfo: LuaLibModulesInfo | undefined;
export function getLuaLibModulesInfo(emitHost: EmitHost): LuaLibModulesInfo {
    if (luaLibModulesInfo === undefined) {
        const lualibPath = path.resolve(__dirname, `../dist/lualib/${luaLibModulesInfoFileName}`);
        const result = emitHost.readFile(lualibPath);
        if (result !== undefined) {
            luaLibModulesInfo = JSON.parse(result) as LuaLibModulesInfo;
        } else {
            throw new Error(`Could not load lualib dependencies from '${lualibPath}'`);
        }
    }
    return luaLibModulesInfo;
}

export function readLuaLibFeature(feature: LuaLibFeature, emitHost: EmitHost): string {
    const featurePath = path.resolve(__dirname, `../dist/lualib/${feature}.lua`);
    const luaLibFeature = emitHost.readFile(featurePath);
    if (luaLibFeature === undefined) {
        throw new Error(`Could not load lualib feature from '${featurePath}'`);
    }
    return luaLibFeature;
}

export function resolveRecursiveLualibFeatures(
    features: Iterable<LuaLibFeature>,
    emitHost: EmitHost,
    luaLibModulesInfo: LuaLibModulesInfo = getLuaLibModulesInfo(emitHost)
): LuaLibFeature[] {
    const loadedFeatures = new Set<LuaLibFeature>();
    const result: LuaLibFeature[] = [];

    function load(feature: LuaLibFeature): void {
        if (loadedFeatures.has(feature)) return;
        loadedFeatures.add(feature);

        const dependencies = luaLibModulesInfo[feature]?.dependencies;
        if (dependencies) {
            dependencies.forEach(load);
        }

        result.push(feature);
    }

    for (const feature of features) {
        load(feature);
    }

    return result;
}

export function loadInlineLualibFeatures(features: Iterable<LuaLibFeature>, emitHost: EmitHost): string {
    let result = "";

    for (const feature of resolveRecursiveLualibFeatures(features, emitHost)) {
        const luaLibFeature = readLuaLibFeature(feature, emitHost);
        result += luaLibFeature + "\n";
    }

    return result;
}

export function loadImportedLualibFeatures(features: Iterable<LuaLibFeature>, emitHost: EmitHost): lua.Statement[] {
    const luaLibModuleInfo = getLuaLibModulesInfo(emitHost);

    const imports = Array.from(features).flatMap(feature => luaLibModuleInfo[feature].exports);

    const requireCall = lua.createCallExpression(lua.createIdentifier("require"), [
        lua.createStringLiteral("lualib_bundle"),
    ]);
    if (imports.length === 0) {
        return [];
    }

    const luaLibId = lua.createIdentifier("____lualib");
    const importStatement = lua.createVariableDeclarationStatement(luaLibId, requireCall);
    const statements: lua.Statement[] = [importStatement];
    // local <export> = ____luaLib.<export>
    for (const item of imports) {
        statements.push(
            lua.createVariableDeclarationStatement(
                lua.createIdentifier(item),
                lua.createTableIndexExpression(luaLibId, lua.createStringLiteral(item))
            )
        );
    }
    return statements;
}

let luaLibBundleContent: string;
export function getLuaLibBundle(emitHost: EmitHost): string {
    if (luaLibBundleContent === undefined) {
        const lualibPath = path.resolve(__dirname, "../dist/lualib/lualib_bundle.lua");
        const result = emitHost.readFile(lualibPath);
        if (result !== undefined) {
            luaLibBundleContent = result;
        } else {
            throw new Error(`Could not load lualib bundle from '${lualibPath}'`);
        }
    }

    return luaLibBundleContent;
}
