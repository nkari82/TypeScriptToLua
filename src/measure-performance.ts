import { performance } from "perf_hooks";

// We use our own performance hooks implementation for easier use, but also call node's performance hooks, so it shows up in the profiler.

let enabled = false;
const marks = new Map<string, number>();
const durations = new Map<string, number>();

function timestamp() {
    return performance.now();
}

/**
 * Marks a performance event, with the given markName.
 */
function mark(markName: string) {
    if (enabled) {
        marks.set(markName, timestamp());
        performance.mark(markName);
    }
}

/**
 * Adds a performance measurement with the specified name.
 *
 * @param measureName The name of the performance measurement.
 * @param startMarkName The name of the starting mark
 * @param endMarkName The name of the ending mark
 */
function measure(measureName: string, startMarkName: string, endMarkName: string) {
    if (enabled) {
        const end = marks.get(endMarkName) ?? timestamp();
        const start = marks.get(startMarkName) ?? performance.timeOrigin;
        const previousDuration = durations.get(measureName) ?? 0;
        durations.set(measureName, previousDuration + (end - start));
        performance.measure(measureName, startMarkName, endMarkName);
    }
}

/**
 * Starts a performance measurement section.
 * @param name name of the measurement
 */
export function startSection(name: string) {
    mark("start " + name);
}

/**
 * Ends a performance measurement section.
 * @param name name of the measurement
 */
export function endSection(name: string) {
    mark("end " + name);
    measure(name, "start " + name, "end " + name);
}

export function isMeasurementEnabled() {
    return enabled;
}

export function enableMeasurement() {
    if (!enabled) {
        enabled = true;
    }
}

export function disableMeasurement() {
    if (enabled) {
        enabled = false;
        marks.clear();
        durations.clear();
    }
}

export function forEachMeasure(callback: (measureName: string, duration: number) => void) {
    durations.forEach((duration, measureName) => callback(measureName, duration));
}

export function getTotalDuration() {
    let total = 0;
    forEachMeasure((_, duration) => (total += duration));
    return total;
}
