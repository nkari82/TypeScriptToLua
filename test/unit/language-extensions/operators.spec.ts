import * as path from "path";
import * as util from "../../util";
import * as tstl from "../../../src";
import { LuaTarget } from "../../../src";
import { unsupportedForTarget, invalidCallExtensionUse } from "../../../src/transformation/utils/diagnostics";

const operatorsProjectOptions: tstl.CompilerOptions = {
    luaTarget: LuaTarget.Lua54,
    types: [path.resolve(__dirname, "../../../language-extensions")],
};

const operatorMapTestObject = `
    class OpTest {
        public value: number;
        public constructor(value: number) { this.value = value; }

        public __add(other: OpTest): OpTest;
        public __add(other: number): number;
        public __add(other: OpTest | number) {
            if (typeof other === "number") {
                return this.value + other;
            } else {
                return new OpTest(this.value + other.value);
            }
        }
        public __sub(other: OpTest) { return new OpTest(this.value - other.value); }
        public __mul(other: OpTest) { return new OpTest(this.value * other.value); }
        public __div(other: OpTest) { return new OpTest(this.value / other.value); }
        public __mod(other: OpTest) { return new OpTest(this.value % other.value); }
        public __pow(other: OpTest) { return new OpTest(this.value ** other.value); }
        public __idiv(other: OpTest) { return new OpTest(Math.floor(this.value / other.value)); }
        public __band(other: OpTest) { return new OpTest(this.value & other.value); }
        public __bor(other: OpTest) { return new OpTest(this.value | other.value); }
        public __bxor(other: OpTest) { return new OpTest(this.value ^ other.value); }
        public __shl(other: OpTest) { return new OpTest(this.value << other.value); }
        public __shr(other: OpTest) { return new OpTest(this.value >>> other.value); }

        public __lt(other: OpTest) { return this.value < other.value; }
        public __gt(other: OpTest) { return this.value > other.value; }

        public __concat(other: OpTest) { return this.value.toString() + other.value.toString(); }

        public __unm() { return -this.value; }
        public __bnot() { return ~this.value; }
        public __len() { return this.value; }
    }
`;

const binaryMathOperatorTests: Array<{ opType: string; left: number; right: number; expectResult: number }> = [
    { opType: "LuaAddition", left: 7, right: 4, expectResult: 11 },
    { opType: "LuaSubtraction", left: 7, right: 4, expectResult: 3 },
    { opType: "LuaMultiplication", left: 7, right: 4, expectResult: 28 },
    { opType: "LuaDivision", left: 7, right: 4, expectResult: 1.75 },
    { opType: "LuaModulo", left: 7, right: 4, expectResult: 3 },
    { opType: "LuaPower", left: 7, right: 4, expectResult: 2401 },
    { opType: "LuaFloorDivision", left: 7, right: 4, expectResult: 1 },
    { opType: "LuaBitwiseAnd", left: 6, right: 5, expectResult: 4 },
    { opType: "LuaBitwiseOr", left: 6, right: 5, expectResult: 7 },
    { opType: "LuaBitwiseExclusiveOr", left: 6, right: 5, expectResult: 3 },
    { opType: "LuaBitwiseLeftShift", left: 7, right: 2, expectResult: 28 },
    { opType: "LuaBitwiseRightShift", left: 7, right: 2, expectResult: 1 },
];

test.each(binaryMathOperatorTests)(
    "binary math operator mapping - global function (%s)",
    ({ opType, left, right, expectResult }) => {
        util.testModule`
        ${operatorMapTestObject}
        declare const op: ${opType}<OpTest, OpTest, OpTest>;
        const left = new OpTest(${left});
        const right = new OpTest(${right});
        export const result = op(left, right).value;
    `
            .setOptions(operatorsProjectOptions)
            .setReturnExport("result")
            .expectToEqual(expectResult);
    }
);

test.each(binaryMathOperatorTests)(
    "binary math operator mapping - static function (%s)",
    ({ opType, left, right, expectResult }) => {
        util.testModule`
        ${operatorMapTestObject}
        declare namespace OpTest {
            export const op: ${opType}<OpTest, OpTest, OpTest>;
        }
        const left = new OpTest(${left});
        const right = new OpTest(${right});
        export const result = OpTest.op(left, right).value;
    `
            .setOptions(operatorsProjectOptions)
            .setReturnExport("result")
            .expectToEqual(expectResult);
    }
);

test.each(binaryMathOperatorTests)(
    "binary math operator mapping - method (%s)",
    ({ opType, left, right, expectResult }) => {
        util.testModule`
        ${operatorMapTestObject}
        declare interface OpTest {
            op: ${opType}Method<OpTest, OpTest>;
        }
        const left = new OpTest(${left});
        const right = new OpTest(${right});
        export const result = left.op(right).value;
    `
            .setOptions(operatorsProjectOptions)
            .setReturnExport("result")
            .expectToEqual(expectResult);
    }
);

const luaTargetsPre53 = [LuaTarget.Lua50, LuaTarget.Lua51, LuaTarget.Lua52, LuaTarget.LuaJIT, LuaTarget.Universal];

const operatorTypesPost53 = [
    "LuaFloorDivision",
    "LuaBitwiseAnd",
    "LuaBitwiseOr",
    "LuaBitwiseExclusiveOr",
    "LuaBitwiseLeftShift",
    "LuaBitwiseRightShift",
];

test.each(luaTargetsPre53.flatMap(target => operatorTypesPost53.map(opType => [target, opType] as const)))(
    "unsupported binary operator mapping (%s %s)",
    (luaTarget, opType) => {
        util.testModule`
        declare interface OpTest { value: number; }
        declare const op: ${opType}<OpTest, OpTest, OpTest>;
        declare const left: OpTest;
        declare const right: OpTest;
        op(left, right);
    `
            .setOptions({ ...operatorsProjectOptions, luaTarget })
            .expectDiagnosticsToMatchSnapshot([unsupportedForTarget.code]);
    }
);

const comparisonOperatorTests: Array<{ opType: string; left: number; right: number; expectResult: boolean }> = [
    { opType: "LuaLessThan", left: 7, right: 4, expectResult: false },
    { opType: "LuaLessThan", left: 4, right: 7, expectResult: true },
    { opType: "LuaGreaterThan", left: 7, right: 4, expectResult: true },
    { opType: "LuaGreaterThan", left: 4, right: 7, expectResult: false },
];

test.each(comparisonOperatorTests)(
    "comparison operator mapping - global function (%s)",
    ({ opType, left, right, expectResult }) => {
        util.testModule`
        ${operatorMapTestObject}
        declare const op: ${opType}<OpTest, OpTest, boolean>;
        const left = new OpTest(${left});
        const right = new OpTest(${right});
        export const result = op(left, right);
    `
            .setOptions(operatorsProjectOptions)
            .setReturnExport("result")
            .expectToEqual(expectResult);
    }
);

test.each(comparisonOperatorTests)(
    "comparison operator mapping - static function (%s)",
    ({ opType, left, right, expectResult }) => {
        util.testModule`
        ${operatorMapTestObject}
        declare namespace OpTest {
            export const op: ${opType}<OpTest, OpTest, boolean>;
        }
        const left = new OpTest(${left});
        const right = new OpTest(${right});
        export const result = OpTest.op(left, right);
    `
            .setOptions(operatorsProjectOptions)
            .setReturnExport("result")
            .expectToEqual(expectResult);
    }
);

test.each(comparisonOperatorTests)(
    "comparison operator mapping - method (%s)",
    ({ opType, left, right, expectResult }) => {
        util.testModule`
        ${operatorMapTestObject}
        declare interface OpTest {
            op: ${opType}Method<OpTest, boolean>;
        }
        const left = new OpTest(${left});
        const right = new OpTest(${right});
        export const result = left.op(right);
    `
            .setOptions(operatorsProjectOptions)
            .setReturnExport("result")
            .expectToEqual(expectResult);
    }
);

test("concat operator mapping - global function", () => {
    util.testModule`
        ${operatorMapTestObject}
        declare const op: LuaConcat<OpTest, OpTest, string>;
        const left = new OpTest(7);
        const right = new OpTest(4);
        export const result = op(left, right);
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual("74");
});

test("concat operator mapping - static function", () => {
    util.testModule`
        ${operatorMapTestObject}
        declare namespace OpTest {
            export const op: LuaConcat<OpTest, OpTest, string>;
        }
        const left = new OpTest(7);
        const right = new OpTest(4);
        export const result = OpTest.op(left, right);
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual("74");
});

test("concat operator mapping - method", () => {
    util.testModule`
        ${operatorMapTestObject}
        declare interface OpTest {
            op: LuaConcatMethod<OpTest, string>;
        }
        const left = new OpTest(7);
        const right = new OpTest(4);
        export const result = left.op(right);
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual("74");
});

const unaryOperatorTests: Array<{ opType: string; operand: number; expectResult: number }> = [
    { opType: "LuaNegation", operand: 13, expectResult: -13 },
    { opType: "LuaBitwiseNot", operand: 13, expectResult: -14 },
    { opType: "LuaLength", operand: 13, expectResult: 13 },
];

test.each(unaryOperatorTests)("unary operator mapping - global function (%s)", ({ opType, operand, expectResult }) => {
    util.testModule`
        ${operatorMapTestObject}
        declare const op: ${opType}<OpTest, number>;
        const operand = new OpTest(${operand});
        export const result = op(operand);
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual(expectResult);
});

test.each(unaryOperatorTests)("unary operator mapping - static function (%s)", ({ opType, operand, expectResult }) => {
    util.testModule`
        ${operatorMapTestObject}
        declare namespace OpTest {
            export const op: ${opType}<OpTest, number>;
        }
        const operand = new OpTest(${operand});
        export const result = OpTest.op(operand);
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual(expectResult);
});

test.each(unaryOperatorTests)("unary operator mapping - method (%s)", ({ opType, operand, expectResult }) => {
    util.testModule`
        ${operatorMapTestObject}
        declare interface OpTest {
            op: ${opType}Method<number>;
        }
        const operand = new OpTest(${operand});
        export const result = operand.op();
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual(expectResult);
});

test.each(luaTargetsPre53)("unsupported unary operator mapping (%s LuaBitwiseNot)", luaTarget => {
    util.testModule`
        declare interface OpTest { value: number; }
        declare const op: LuaBitwiseNot<OpTest, OpTest>;
        declare const operand: OpTest;
        op(operand);
    `
        .setOptions({ ...operatorsProjectOptions, luaTarget })
        .expectDiagnosticsToMatchSnapshot([unsupportedForTarget.code]);
});

test("operator mapping overload - global function", () => {
    util.testModule`
        ${operatorMapTestObject}
        declare const op: LuaAddition<OpTest, OpTest, OpTest> & LuaAddition<OpTest, number, number>;
        const left = new OpTest(4);
        const right = new OpTest(7);
        const resultA = op(left, right).value;
        const resultB = op(left, 13);
        export const result = {resultA, resultB};
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual({ resultA: 11, resultB: 17 });
});

test("operator mapping overload - static function", () => {
    util.testModule`
        ${operatorMapTestObject}
        declare namespace OpTest {
            export const op: LuaAddition<OpTest, OpTest, OpTest> & LuaAddition<OpTest, number, number>;
        }
        const left = new OpTest(4);
        const right = new OpTest(7);
        const resultA = OpTest.op(left, right).value;
        const resultB = OpTest.op(left, 13);
        export const result = {resultA, resultB};
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual({ resultA: 11, resultB: 17 });
});

test("operator mapping overload - method", () => {
    util.testModule`
        ${operatorMapTestObject}
        declare interface OpTest {
            op: LuaAdditionMethod<OpTest, OpTest> & LuaAdditionMethod<number, number>;
        }
        const left = new OpTest(4);
        const right = new OpTest(7);
        const resultA = left.op(right).value;
        const resultB = left.op(13);
        export const result = {resultA, resultB};
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual({ resultA: 11, resultB: 17 });
});

test("operator mapping - method element call", () => {
    util.testModule`
        ${operatorMapTestObject}
        declare interface OpTest {
            add: LuaAdditionMethod<number, number>;
        }
        const left = new OpTest(7);
        export const result = left["add"](4);
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual(11);
});

test("operator mapping - function element call", () => {
    util.testModule`
        ${operatorMapTestObject}
        declare namespace OpTest {
            export const add: LuaAddition<OpTest, number, number>;
        }
        const left = new OpTest(7);
        export const result = OpTest["add"](left, 4);
    `
        .setOptions(operatorsProjectOptions)
        .setReturnExport("result")
        .expectToEqual(11);
});

test.each([
    "const foo: unknown = op;",
    "const foo = `${op}`",
    "declare function foo(op: LuaAddition<number, number, number>): void; foo(op);",
    "const foo = (op as any)(1, 2);",
    "const foo = [op];",
])("operator mapping - invalid use (%s)", invalidCode => {
    util.testModule`
        declare const op: LuaAddition<number, number, number>;
        ${invalidCode}
    `
        .setOptions(operatorsProjectOptions)
        .expectDiagnosticsToMatchSnapshot([invalidCallExtensionUse.code]);
});
