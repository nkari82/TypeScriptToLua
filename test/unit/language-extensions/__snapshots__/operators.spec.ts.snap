// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operator mapping - invalid use (const foo = (op as any)(1, 2);): code 1`] = `"foo = op(_G, 1, 2)"`;

exports[`operator mapping - invalid use (const foo = (op as any)(1, 2);): diagnostics 1`] = `"main.ts(3,22): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`operator mapping - invalid use (const foo = [op];): code 1`] = `"foo = {op}"`;

exports[`operator mapping - invalid use (const foo = [op];): diagnostics 1`] = `"main.ts(3,22): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`operator mapping - invalid use (const foo = \`\${op}\`): code 1`] = `"foo = tostring(op)"`;

exports[`operator mapping - invalid use (const foo = \`\${op}\`): diagnostics 1`] = `"main.ts(3,24): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`operator mapping - invalid use (const foo: unknown = op;): code 1`] = `"foo = op"`;

exports[`operator mapping - invalid use (const foo: unknown = op;): diagnostics 1`] = `"main.ts(3,30): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`operator mapping - invalid use (declare function foo(op: LuaAddition<number, number, number>): void; foo(op);): code 1`] = `"foo(_G, op)"`;

exports[`operator mapping - invalid use (declare function foo(op: LuaAddition<number, number, number>): void; foo(op);): diagnostics 1`] = `"main.ts(3,82): error TSTL: This function must be called directly and cannot be referred to."`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseAnd): code 1`] = `"local ____ = left & right"`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseAnd): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseExclusiveOr): code 1`] = `"local ____ = left ~ right"`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseExclusiveOr): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseLeftShift): code 1`] = `"local ____ = left << right"`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseLeftShift): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseOr): code 1`] = `"local ____ = left | right"`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseOr): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseRightShift): code 1`] = `"local ____ = left >> right"`;

exports[`unsupported binary operator mapping (5.1 LuaBitwiseRightShift): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (5.1 LuaFloorDivision): code 1`] = `"local ____ = left // right"`;

exports[`unsupported binary operator mapping (5.1 LuaFloorDivision): diagnostics 1`] = `"main.ts(6,9): error TSTL: Floor division operator is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseAnd): code 1`] = `"local ____ = left & right"`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseAnd): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.2."`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseExclusiveOr): code 1`] = `"local ____ = left ~ right"`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseExclusiveOr): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.2."`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseLeftShift): code 1`] = `"local ____ = left << right"`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseLeftShift): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.2."`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseOr): code 1`] = `"local ____ = left | right"`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseOr): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.2."`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseRightShift): code 1`] = `"local ____ = left >> right"`;

exports[`unsupported binary operator mapping (5.2 LuaBitwiseRightShift): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.2."`;

exports[`unsupported binary operator mapping (5.2 LuaFloorDivision): code 1`] = `"local ____ = left // right"`;

exports[`unsupported binary operator mapping (5.2 LuaFloorDivision): diagnostics 1`] = `"main.ts(6,9): error TSTL: Floor division operator is/are not supported for target Lua 5.2."`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseAnd): code 1`] = `"local ____ = left & right"`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseAnd): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target LuaJIT."`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseExclusiveOr): code 1`] = `"local ____ = left ~ right"`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseExclusiveOr): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target LuaJIT."`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseLeftShift): code 1`] = `"local ____ = left << right"`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseLeftShift): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target LuaJIT."`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseOr): code 1`] = `"local ____ = left | right"`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseOr): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target LuaJIT."`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseRightShift): code 1`] = `"local ____ = left >> right"`;

exports[`unsupported binary operator mapping (JIT LuaBitwiseRightShift): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target LuaJIT."`;

exports[`unsupported binary operator mapping (JIT LuaFloorDivision): code 1`] = `"local ____ = left // right"`;

exports[`unsupported binary operator mapping (JIT LuaFloorDivision): diagnostics 1`] = `"main.ts(6,9): error TSTL: Floor division operator is/are not supported for target LuaJIT."`;

exports[`unsupported binary operator mapping (universal LuaBitwiseAnd): code 1`] = `"local ____ = left & right"`;

exports[`unsupported binary operator mapping (universal LuaBitwiseAnd): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (universal LuaBitwiseExclusiveOr): code 1`] = `"local ____ = left ~ right"`;

exports[`unsupported binary operator mapping (universal LuaBitwiseExclusiveOr): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (universal LuaBitwiseLeftShift): code 1`] = `"local ____ = left << right"`;

exports[`unsupported binary operator mapping (universal LuaBitwiseLeftShift): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (universal LuaBitwiseOr): code 1`] = `"local ____ = left | right"`;

exports[`unsupported binary operator mapping (universal LuaBitwiseOr): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (universal LuaBitwiseRightShift): code 1`] = `"local ____ = left >> right"`;

exports[`unsupported binary operator mapping (universal LuaBitwiseRightShift): diagnostics 1`] = `"main.ts(6,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported binary operator mapping (universal LuaFloorDivision): code 1`] = `"local ____ = left // right"`;

exports[`unsupported binary operator mapping (universal LuaFloorDivision): diagnostics 1`] = `"main.ts(6,9): error TSTL: Floor division operator is/are not supported for target Lua 5.1."`;

exports[`unsupported unary operator mapping (5.1 LuaBitwiseNot): code 1`] = `"local ____ = ~operand"`;

exports[`unsupported unary operator mapping (5.1 LuaBitwiseNot): diagnostics 1`] = `"main.ts(5,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;

exports[`unsupported unary operator mapping (5.2 LuaBitwiseNot): code 1`] = `"local ____ = ~operand"`;

exports[`unsupported unary operator mapping (5.2 LuaBitwiseNot): diagnostics 1`] = `"main.ts(5,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.2."`;

exports[`unsupported unary operator mapping (JIT LuaBitwiseNot): code 1`] = `"local ____ = ~operand"`;

exports[`unsupported unary operator mapping (JIT LuaBitwiseNot): diagnostics 1`] = `"main.ts(5,9): error TSTL: Native bitwise operations is/are not supported for target LuaJIT."`;

exports[`unsupported unary operator mapping (universal LuaBitwiseNot): code 1`] = `"local ____ = ~operand"`;

exports[`unsupported unary operator mapping (universal LuaBitwiseNot): diagnostics 1`] = `"main.ts(5,9): error TSTL: Native bitwise operations is/are not supported for target Lua 5.1."`;
